server/src/version_provider.ts

```diff
@@ -6,65 +6,134 @@
+      paths: [location],
+    });
+    const packageJson = require(packageJsonPath);
+    const resolvedPath = require.resolve(packageName, {
+      paths: [location],
+    });
     return {
-      resolvedPath: path.dirname(resolvedPath),
-      version: packageJson.version,
+      resolvedPath,
+      version: new Version(packageJson.version),
     };
   } catch {
   }
 }
 
-function minVersion(nodeModule: NodeModule, minMajor: number): boolean {
-  if (!nodeModule.version) {
-    return false;
-  }
-  const [majorStr] = nodeModule.version.split('.');
-  if (!majorStr) {
-    return false;
+/**
+ * Resolve the node module with the specified `packageName` that satisfies
+ * the specified minimum version.
+ * @param packageName name of package to be resolved
+ * @param minVersionStr minimum version
+ * @param probeLocations locations to initiate node module resolution
+ * @param rootPackage location of package.json, if different from `packageName`
+ */
+function resolveWithMinVersion(
+    packageName: string, minVersionStr: string, probeLocations: string[],
+    rootPackage?: string): NodeModule {
+  if (rootPackage && !packageName.startsWith(rootPackage)) {
+    throw new Error(`${packageName} must be in the root package`);
   }
-  const major = Number(majorStr);
-  if (isNaN(major)) {
-    return false;
+  const minVersion = new Version(minVersionStr);
+  for (const location of probeLocations) {
+    const nodeModule = resolve(packageName, location, rootPackage);
+    if (nodeModule && nodeModule.version.greaterThanOrEqual(minVersion)) {
+      return nodeModule;
+    }
   }
-  return major >= minMajor;
+  throw new Error(
+      `Failed to resolve '${packageName}' with minimum version '${minVersion}' from ` +
+      JSON.stringify(probeLocations, null, 2));
 }
 
 /**
- * Resolve the node module with the specified `packageName` that satisfies
- * the specified minimum major version.
- * @param packageName
- * @param minMajor
+ * Resolve `typescript/lib/tsserverlibrary` from the given locations.
  * @param probeLocations
  */
-export function resolveWithMinMajor(
-    packageName: string, minMajor: number, probeLocations: string[]): NodeModule {
-  for (const location of probeLocations) {
-    const nodeModule = resolve(packageName, [location]);
-    if (!nodeModule) {
-      continue;
+export function resolveTsServer(probeLocations: string[]): NodeModule {
+  const tsserver = 'typescript/lib/tsserverlibrary';
+  return resolveWithMinVersion(tsserver, MIN_TS_VERSION, probeLocations, 'typescript');
+}
+
+/**
+ * Resolve `@angular/language-service` from the given locations.
+ * @param probeLocations
+ */
+export function resolveNgLangSvc(probeLocations: string[]): NodeModule {
+  return resolveWithMinVersion('@angular/language-service', MIN_NG_VERSION, probeLocations);
+}
+
+/**
+ * Converts the specified string `a` to integer. Returns -1 if the result is NaN.
+ * @param a
+ */
+function atoi(a: string): number {
```

Comment:
can we rename this to `parseNumber` or similar? I'm afraid the C convention may be unfamiliar to people who haven't worked with that

Reply:
renamed to `parseNonNegativeInt`